### ISSUE TITLE
[Wave] Fun projects for beginnners

### ISSUE THREAD

**Issue by raikonenfnu** [iree-org/iree-turbine#278]
Are you interested in learning more about GPU programming and developing cool optimizations? Do you want to help build next generation and state-of-the-art machine learning models and layers? Do you want to define the future programming paradigm of machine learning and GPU layers? Look no further, come join us in building "Wave"!

Here are some fun starter tasks to look at:

- [ ] Implement backwards attention kernel on TKW
- [ ] Cache kernels
- [ ] Scalar support
- [ ] Auto detection of hardware architecture in runtime
- [ ] Support more architecture in codegen aside from CDNA.

**Comment by NoumanAmir657** [comment]
Would this be open for someone who has almost no knowledge of GPU programming except for some basic understanding.

**Comment by raikonenfnu** [comment]
Hi @NoumanAmir657 thanks for reaching out! There are plenty of tasks to go around, depending on your experience level, would probably assign easier tasks to start with.

**Comment by NoumanAmir657** [comment]
> Hi @NoumanAmir657 thanks for reaching out! There are plenty of tasks to go around, depending on your experience level, would probably assign easier tasks to start with.

Hi, yes would love any pointers to which task would help me getting started. You can assign me whichever you see fit and I can work towards it

**Comment by raikonenfnu** [comment]
@NoumanAmir657 I think a good starter task is to implement `tkw.abs` and it's lowering into `math.absfOp` or `math.absIOp` in math dialect of MLIR. Here is a sample of a similar task on reciprocal (https://github.com/iree-org/iree-turbine/commit/82852d13198be9f18f6a070b2762bdaf6e78c5f3).

LMK what you think. :)

Disclaimer: if this(or any task assigned to external collaborator) becomes high priority, someone internal may take on this task to get it out asap.

**Comment by NoumanAmir657** [comment]
> @NoumanAmir657 I think a good starter task is to implement `tkw.abs` and it's lowering into `math.absfOp` or `math.absIOp` in math dialect of MLIR. Here is a sample of a similar task on reciprocal (https://github.com/iree-org/iree-turbine/commit/82852d13198be9f18f6a070b2762bdaf6e78c5f3).
> 
> LMK what you think. :)
> 
> Disclaimer: if this(or any task assigned to external collaborator) becomes high priority, someone internal may take on this task to get it out asap.

Yes this seems a good starting point. I shall make a PR soon. Thanks!

**Comment by raikonenfnu** [comment]
thanks @NoumanAmir657 for the awesome work with tkw.abs! let me know if you're interested in picking up other issues. The next important thing for us is adding support for tkw.minimum (elementwise) and extending it's support to tkw.min(reduction) but feel free to pick other things more aligned with your interest of course.

**Comment by NoumanAmir657** [comment]
> thanks @NoumanAmir657 for the awesome work with tkw.abs! let me know if you're interested in picking up other issues. The next important thing for us is adding support for tkw.minimum (elementwise) and extending it's support to tkw.min(reduction) but feel free to pick other things more aligned with your interest of course.

Thanks for getting me started. I want to contribute more. Over the weekend, I will go over issues and decide on one. Thanks!

**Comment by NoumanAmir657** [comment]
@raikonenfnu Hi. You can assign me the tkw.minimum issue.
Thanks!

**Comment by raikonenfnu** [comment]
> Hi. You can assign me the tkw.minimum issue.

@NoumanAmir657 SG! thanks :)

**Comment by egebeysel** [comment]
Hi, can I volunteer for the `tkw.round_even` support?, thanks :)

**Comment by raikonenfnu** [comment]
> Hi, can I volunteer for the tkw.round_even support?, thanks :)

Sounds great! :)

**Comment by raikonenfnu** [comment]
@egebeysel Here are some sample PRs that does similar thing to what you'd be doing: (https://github.com/iree-org/iree-turbine/commit/82852d13198be9f18f6a070b2762bdaf6e78c5f3, https://github.com/iree-org/iree-turbine/commit/71eb1c8bfbb8c053a3684caabdc7df6648e7d794) 

But in this case, you'd want to lower tkw.round_to_even to math.round_even from the math standard dialect in MLIR https://mlir.llvm.org/docs/Dialects/MathOps/#mathroundeven-mathroundevenop. Do reach out if you have any questions! :)

**Comment by ziereis** [comment]
@raikonenfnu Hello, I have some questions regarding "Support more architectures in codegen aside from CDNA." Currently, the codegen directly emits ops from the AMDGPU dialect, like barriers, MMA ops, etc. What is the plan for supporting GPUs from different vendors? Are these ops going to be turned into a higher-level dialect, for instance a vector.contract instead of a amdgpu.mma and let iree handle the lowering? Or is the plan to directly emit the ops for the target architecture in codegen.py?

**Comment by raikonenfnu** [comment]
Hey @ziereis SFLR, just got back from holidays!

> Are these ops going to be turned into a higher-level dialect, for instance a vector.contract instead of a amdgpu.mma and let iree handle the lowering? Or is the plan to directly emit the ops for the target architecture in codegen.py?

We are planning to keep most of the emitting of ops in iree-turbine (not iree proper). Hence the latter, or a variant of the latter option would most likely be the way we are going down. If you are interested in adding support on a new target architecture, do let me or us know! Would be happy to brainstorm over an GH issue or VC. :)

**Comment by NoumanAmir657** [comment]
@raikonenfnu you can assign me that `tkw.min` issue. Thanks!

**Comment by NoumanAmir657** [comment]
@raikonenfnu Can you help me pick another issue?

**Comment by NoumanAmir657** [comment]
@raikonenfnu I would like to work on the 1st one. Can you point me to how to get started on this?

**Comment by raikonenfnu** [comment]
@NoumanAmir657 Sorry for the late response, here is an issue that gives some general context and guidance on how to get started https://github.com/iree-org/iree-turbine/issues/599. LMK if you have any Qs

P.S feel free to ping if I haven't replied in a few days, work does get hectic here but I wouldn't want your questions or asks to be buried under! :)
